src/compiler/moduleNameResolver.ts

```diff
@@ -2447,11 +2428,23 @@ function loadModuleFromImmediateNodeModulesDirectory(extensions: Extensions, mod
         trace(state.host, Diagnostics.Directory_0_does_not_exist_skipping_all_lookups_in_it, nodeModulesFolder);
     }
 
-    const packageResult = typesScopeOnly ? undefined : loadModuleFromSpecificNodeModulesDirectory(extensions, moduleName, nodeModulesFolder, nodeModulesFolderExists, state, cache, redirectedReference);
-    if (packageResult) {
-        return packageResult;
+    // Search with the following priority:
+    //   1. TS/DTS files in the implementation package
+    //   2. DTS files in the @types package
+    //   3. JS files in the implementation package
+    const priorityExtensions = extensions & (Extensions.TypeScript | Extensions.Declaration);
+    const secondaryExtensions = extensions & ~(Extensions.TypeScript | Extensions.Declaration);

```

Comment:
The primary and secondary names here with the comments is confusing since they correspond to points 1 and 3
I kept thinking how is secondary same as option 2 and then when i saw usage it became clear that its actually point 3,